pmount.1
.TH pmount 1 "August 27, 2004" "Martin Pitt"

.SH NAME
pmount \- mount arbitrary hotpluggable devices as normal user

.SH SYNOPSIS

.B pmount
[
.I options
]
.I device

.B pmount
[
.I options
]
.I device label

.B pmount \-\-lock
[
.I options
]
.I device pid

.B pmount \-\-unlock
[
.I options
]
.I device pid

.SH DESCRIPTION

pmount ("policy mount") is a wrapper around the standard mount program which
permits normal users to mount removable devices without a matching /etc/fstab
entry.

pmount also supports encrypted devices which use dm-crypt and have
LUKS metadata. If a LUKS-capable
.B cryptsetup
is installed, pmount will use it to decrypt the device first and mount
the mapped unencrypted device instead.

.B pmount
is invoked like this:

.B pmount
.I device 
[
.I label
]

This will mount 
.I device
to a directory below /media if policy is met (see below). If 
.I label
is given, the mount point will be /media/label, otherwise it will be
/media/device.

The device will be mounted with the following flags: 
sync,atime,nodev,noexec,noauto,nosuid,user,rw

Some applications like CD burners modify a raw device which must not be mounted
while the burning process is in progress. To prevent automatic mounting, pmount
offers a locking mechanism:
.B pmount \-\-lock 
.I device pid
will prevent the pmounting of 
.I device
until it is unlocked again using 
.B pmount \-\-unlock 
.I device pid\fR. The process id 
.I pid
assigns the lock to a particular process; this allows to lock a device by
several processes.

During mount, the list of locks is cleaned, i. e. all locks whose associated
process does not exist any more are removed. This prevents forgotten indefinite
locks from crashed programs.

.SH POLICY

The mount will succeed if all of the following conditions are met:

.IP \(bu 2
.I device
is a block device in /dev/
.IP \(bu 
.I device
is not in /etc/fstab (if it is, pmount executes \fB mount \fI device\fR as the
calling user to handle this transparently)
.IP \(bu
.I device
is not already mounted according to /etc/mtab and /proc/mounts
.IP \(bu
if the mount point already exists, there is no device already mounted at it
and the directory is empty 
.IP \(bu
.I device
is removable (USB or FireWire device, or \fB/sys/block/\fIdrive\fB/removable\fR
is 1) or whitelisted in
.B /etc/pmount.allow\fR.
.IP \(bu
.I device
is not locked

.SH OPTIONS

.TP
.B \-r, \-\-read-only
Force the device to be mounted read only. If neither \-r nor \-w is
specified, the kernel will choose an appropriate default.

.TP
.B \-w, \-\-read-write
Force the device to be mounted read/write. If neither \-r nor \-w is
specified, the kernel will choose an appropriate default.

.TP
.B \-s, \-\-sync
Mount the device with the 
.B sync
option, i. e. without write caching. Default is 
.B async
(write-back). With this option, write operations are much slower and
due to the massive increase of updates of inode/FAT structures, flash
devices may suffer heavily if you write large files. This option is
intended to make it safe to just rip out USB drives without proper
unmounting.

.TP
.B \-A, \-\-noatime
Mount the device with the 
.B noatime
option. Default is 
.B atime\fR. 

.TP
.B \-e, \-\-exec
Mount the device with the 
.B exec
option. Default is 
.B noexec\fR. 

.TP
.B \-t \fIfilesystem\fR, \fB\-\-type \fIfilesystem
Mount as specified file system type. The file system type is automatically
determined if this option is not given.

.TP
.B \-c \fIcharset\fR, \fB\-\-charset \fIcharset
Use given I/O character set (default: 
.B utf8 
if called in an UTF-8 locale, otherwise mount default). This
corresponds with the mount option
.B iocharset\fR. This option is ignored for file systems that do not
support setting the character set (see
.B mount
(8) for details).

.TP
.B \-u \fIumask\fR, \fB\-\-umask \fIumask\fR
Use specified umask instead of the default one. For UDF, the default
is '007', for VFAT and NTFS the default is '077'. This value is
ignored for file systems which do not support setting an umask.

.TP
.B \-\-passphrase \fIfile
If the device is encrypted (dm-crypt with LUKS metadata), read the
passphrase from specified
.I file
instead of prompting at the terminal.

.TP
.B \-h, \-\-help
Print a help message and exit successfully.

.TP
.B \-d, \-\-debug
Enable verbose debug messages.

.SH FILES

.TP
.B /etc/pmount.allow
List of devices (one device per line) which are additionally permitted
for pmounting.

.SH SEE ALSO

.BR pumount (1),
.BR mount (8)

.SH AUTHOR
.B pmount 
is developed by Martin Pitt <martin.pitt@canonical.com>.


CHANGES
Development
-----------
- pmount-hal: support device argument
- pmount manpage: fix default mount options inconsistency
- pmount-hal: full support for HAL expored mount policy (including
  global and per-drive policy, and the "exec" option)

0.9.5
-----
- pmount: Fix forgotten line break in debug output
- pmount: improve debugging of removability detection
- add README.devel with some VCS explanations
- pmount: atomically lock the mount point to avoid races with parallel
  pmount instances, and thus double-mounts to the same directory
- pmount-hal: gracefully handle pmount failures due to locked mount
  points
- pmount-hal: convert to use libhal-storage for better abstraction
- pmount-hal: if volume file system does not exist, fall back to
  storage file system
- pmount: Make read-only encrypted devices work; thanks to Sören
  Köpping for the patch
- updated translations with latest Rosetta data

0.9.4
-----
- policy.c: print out meaningful error message if compiling the
  regular expression fails
- policy.c: fix regular expression to work with all locales

0.9.3
-----
- pmount: close an already opened LUKS device if mount failed
- pmount: print a proper error message if mapped device already exists
- pmount: do not explicitly mount with "rw" or "ro" options since the
  default usually does the right thing
- pmount: add -r and -w switches

0.9.2
-----
- Switched version control to bazaar-ng (http://www.bazaar-ng.org);
  publicly accessible branch is at
  http://people.ubuntu.com/~pitti/bzr/pmount/
- pmount-hal: ignore invalid file systems, autodetect in this case
- Mark JFS and iso9660 as iocharset-capable
- pmount-hal: read volume.policy.mount_option.umask from hal.
- Fix "make dist" to work properly with the bazaar-ng metadata

0.9.1
-----
- Makefile: fix "make dist" tarball generation
- pmount: use async by default to not destroy flash memory

0.9
---
- pmount-hal: ported to new hal 0.5 and dbus 0.32 API
- integrate cryptsetup (with LUKS extension) if device is encrypted
  with dmcrypt and LUKS metadata
- pmount: mount vfat with shortname=winnt option
- pmount-hal: use volume.policy.mount_filesystem instead of
  volume.fstype
- always build the POT file, delete it on "make clean"
- Makefile: added "dist" target

0.8
---
- unbreak locking, this created ".created_by_pmount" stamps in the
  locking directories
- accept pumount by label name
- accept pmount by mount path (only in /media)
- added --umask option
- renamed template.pot to pmount.pot to match translation domain
- support a whitelist in /etc/pmount.allow

0.7.2
-----
- updated translations from Rosetta: ca, es_ES
- added new translations from Rosetta: cs, es, fr, hr, it_IT, nb

0.7.1
-----
- pmount-hal: consider a directory empty if it only contains a
  .created-by-pmount file; this makes it work with unmounting using
  umount instead of pumount

0.7
---
- support --exec/-e option
- rewrote pmount-hal in C, now it runs 10 times faster

0.6
---
- pmount-hal: use hal-supplied file system type (if present) to speed up
  mounting and to avoid bogus error messages in kernel log
- mount vfat with 'quiet' option
- support pmount/pumount'ing mount points
- added Catalan translation, thanks to Eduard Duran
- added Spanish translation, thanks to Gustavo de Lama and Jorge
  González González

0.5.1
-----
- pmount: do not allow empty media labels; this allowed to mount a
  device directly under /media if /media was empty

0.5
---
- use stamp files to mark directories created by pmount and only
  remove a mount point directory if it contains a stamp file
- use the given device name as default mount point name instead of the
  symlink-resolved one
- i18n support
- added German translation

0.4.4
-----
- pmount-hal: resolve symlinks before checking mount point directories
- use umask "077" instead of "007" for VFAT and NTFS volumes; this allows e. g.
  to use ssh keys on USB sticks
- if a mount with given iocharset fails, try again without; there were reports
  of failed vfat mounts when giving iocharset

0.4.3
-----
- pmount-hal: when checking for already existing mount points, also
  accept existing, but empty directories (if nothing is mounted at
  them)

0.4.2
-----
- pumount: allow to unmount any device if called as root


src/pmount.c
/**
 * pmount.c - policy wrapper around 'mount' to allow mounting removable devices
 *            for normal users
 *
 * Author: Martin Pitt <martin.pitt@canonical.com>
 * (c) 2004 Canonical Ltd.
 * 
 * This software is distributed under the terms and conditions of the 
 * GNU General Public License. See file GPL for the full text of the license.
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <limits.h>
#include <unistd.h>
#include <fcntl.h>
#include <dirent.h>
#include <getopt.h>
#include <errno.h>
#include <locale.h>
#include <langinfo.h>
#include <libintl.h>

#include "fs.h"
#include "policy.h"
#include "utils.h"
#include "luks.h"

/* error codes */
const int E_ARGS = 1;
const int E_DEVICE = 2;
const int E_MNTPT = 3;
const int E_POLICY = 4;
const int E_EXECMOUNT = 5;
const int E_UNLOCK = 6;
const int E_PID = 7;
const int E_LOCKED = 8;
const int E_INTERNAL = 100;

/**
 * Print some help.
 * @param exename Name of the executable (argv[0]).
 */
void
usage( const char* exename )
{
    printf( _("Usage:\n\n%s [options] <device> [<label>]\n\n"
    "  Mount <device> to a directory below %s if policy requirements\n"
    "  are met (see pmount(1) for details). If <label> is given, the mount point\n"
    "  will be %s/<label>, otherwise it will be %s<device>.\n"
    "  If the mount point does not exist, it will be created.\n\n"),
        exename, MEDIADIR, MEDIADIR, MEDIADIR );

    printf( _("%s --lock <device> <pid>\n"
    "  Prevent further pmounts of <device> until it is unlocked again. <pid>\n"
    "  specifies the process id the lock holds for. This allows to lock a device\n"
    "  by several independent processes and avoids indefinite locks of crashed\n"
    "  processes (nonexistant pids are cleaned before attempting a mount).\n\n"),
        exename );

    printf( _("%s --unlock <device> <pid>\n"
    "  Remove the lock on <device> for process <pid> again.\n\n"),
        exename);
    puts( _("Options:\n"
    "  -r          : force <device> to be mounted read-only\n"
    "  -w          : force <device> to be mounted read-write\n"
    "  -s, --sync  : mount <device> with the 'sync' option (default: 'async')\n"
    "  --noatime   : mount <device> with the 'noatime' option (default: 'atime')\n"
    "  -e, --exec  : mount <device> with the 'exec' option (default: 'noexec')\n"
    "  -t <fs>     : mount as file system type <fs> (default: autodetected)\n"
    "  -c <charset>: use given I/O character set (default: 'utf8' if called\n"
    "                in an UTF-8 locale, otherwise mount default)\n"
    "  -u <umask>  : use specified umask instead of the default (only for\n"
    "                file sytems which actually support umask setting)\n"
    " --passphrase <file>\n"
    "                read passphrase from file instead of the terminal\n"
    "                (only for LUKS encrypted devices)\n"
    "  -d, --debug : enable debug output (very verbose)\n"
    "  -h, --help  : print help message and exit successfuly") );
}

/**
 * Check whether the user is allowed to mount the given device to the given
 * mount point. Creates the mount point if it does not exist yet. 
 * @return 0 on success, -1 on failure
 */
int
check_mount_policy( const char* device, const char* mntpt ) 
{
    int result = device_valid( device ) &&
        !device_mounted( device, 0, NULL ) &&
        ( device_whitelisted( device ) || device_removable( device ) ) &&
        !device_locked( device ) &&
        mntpt_valid( mntpt ) &&
        !mntpt_mounted( mntpt, 0 );

    if( result ) 
        debug( "policy check passed\n" ); 
    else 
        debug( "policy check failed\n" );

    /* the policy functions deliver booleans, but we want a standard Unix
       result */
    return result ? 0 : -1;
}

/**
 * Create a mount point pathname.
 * @param device device for which a moint point is created
 * @param label if NULL, the mount point will be MEDIADIR/device, otherwise
 *        MEDIADIR/label
 * @param mntpt buffer to write the mount point pathname to
 * @param mntpt_size size of mntpt in characters
 * @return 0 on success, -1 on failure
 */
int
make_mountpoint_name( const char* device, const char* label, char* mntpt,
        size_t mntpt_size )
{
    char* d;
    int media_dir_len = strlen( MEDIADIR );

    /* does the device start with DEVDIR? */
    if( strncmp( device, DEVDIR, sizeof( DEVDIR )-1 ) ) { 
        fprintf( stderr, _("Error: make_mountpoint_name: invalid device %s (must be in /dev/)\n"), device ); 
        return -1;
    }

    if( label ) {
        /* ignore a leading MEDIADIR */
        if( !strncmp( label, MEDIADIR, media_dir_len ) )
            label += media_dir_len;

        if( !*label ) {
            fprintf( stderr, _("Error: label must not be empty\n") );
            return -1;
        }
        if( strlen( label ) > MAX_LABEL_SIZE ) {
            fprintf( stderr, _("Error: label too long\n") );
            return -1;
        }

        if( strchr( label, '/' ) ) {
            fprintf( stderr, _("Error: '/' must not occur in label name\n") );
            return -1;
        }

        snprintf( mntpt, mntpt_size, "%s%s", MEDIADIR, label );
    } else {
        if( strlen( device ) > MAX_LABEL_SIZE ) {
            fprintf( stderr, _("Error: device name too long\n") );
            return -1;
        }

        /* chop the DEVDIR prefix */
        device += sizeof( DEVDIR )-1;

        /* get rid of slashes */
        d = strreplace( device, '/', '_' );

        snprintf( mntpt, mntpt_size, "%s%s", MEDIADIR, d );
        free( d );
    }

    debug( "mount point to be used: %s\n", mntpt );
    return 0;
}

/**
 * Drop all privileges and exec 'mount device'. Does not return on success, if
 * it returns, MOUNTPROG could not be executed.
 */
void
do_mount_fstab( const char* device )
{
    debug( "device %s handled by fstab, calling mount\n", device );

    /* drop all privileges and transparently call mount */
    get_root();
    if( setuid( getuid() ) ) {
        perror( _("Error: could not drop all uid privileges") );
        return;
    }

    execl( MOUNTPROG, MOUNTPROG, device, NULL );
    perror( _("Error: could not execute mount") );
}

/**
 * Raise to full privileges and call mount with given file system. Exits the
 * program immediately if MOUNTPROG cannot be executed or the given file system
 * is invalid. NOTE: This function must not exit() since it is called in a
 * lock-unlock-block.
 * @param device device node to mount
 * @param mntpt desired mount point
 * @param fsname file system name (mount option -t)
 * @param async if not 0, the device will be mounted with 'async' (i. e. write
 *        caching)
 * @param noatime if not 0, the device will be mounted with 'noatime'
 * @param exec if not 0, the device will be mounted with 'exec'
 * @param force_write 1 for forced r/w, 0 for forced r/o, -1 for kernel default
 * @param iocharset charset to use for file name conversion; NULL for mount
 *        default
 * @param umask User specified umask (NULL for default)
 * @param suppress_errors: if true, stderr is redirected to /dev/null
 * @return exit status of mount, or -1 on failure.
 */
int
do_mount( const char* device, const char* mntpt, const char* fsname, int async,
        int noatime, int exec, int force_write, const char* iocharset, const
        char* umask, int suppress_errors )
{
    const struct FS* fs;
    char ugid_opt[100];
    char umask_opt[100];
    char iocharset_opt[100];
    const char* sync_opt = ",sync";
    const char* atime_opt = ",atime";
    const char* exec_opt = ",noexec";
    const char* access_opt = NULL;
    char options[1000];

    /* check and retrieve option information for requested file system */
    if( !fsname) {
        fprintf( stderr, _("Internal error: mount_attempt: given file system name is NULL\n") );
        return -1;
    }

    fs = get_fs_info( fsname );
    if( !fs ) {
        fprintf( stderr, _("Error: invalid file system name '%s'\n"), fsname );
        return -1;
    }

    /* validate user specified umask */
    if( umask && parse_unsigned( umask, E_ARGS ) > 0777 ) {
        fprintf( stderr, _("Error: invalid umask %s\n"), umask );
        return -1;
    }

    /* assemble option string */
    *ugid_opt = *umask_opt = *iocharset_opt = 0;
    if( fs->support_ugid )
        snprintf( ugid_opt, sizeof( ugid_opt ), ",uid=%i,gid=%i", 
                getuid(), getgid() );

    if( fs->umask )
        snprintf( umask_opt, sizeof( umask_opt ), ",umask=%s", 
                umask ? umask : fs->umask );

    if( async )
        sync_opt = ",async";

    if( noatime )
        atime_opt = ",noatime";

    if( exec )
        exec_opt = ",exec";

    if( force_write == 0 )
        access_opt = ",ro";
    else if( force_write == 1 )
        access_opt = ",rw";
    else
        access_opt = "";

    if( iocharset && fs->support_iocharset ) {
        if( !is_word_str( iocharset ) ) {
            fprintf( stderr, _("Error: invalid charset name '%s'\n"), iocharset );
            return -1;
        }
        snprintf( iocharset_opt, sizeof( iocharset_opt ), ",iocharset=%s", iocharset );
    }

    snprintf( options, sizeof( options ), "%s%s%s%s%s%s%s%s", 
            fs->options, sync_opt, atime_opt, exec_opt, access_opt, ugid_opt,
            umask_opt, iocharset_opt );

    /* go for it */
    return spawnl( SPAWN_EROOT | SPAWN_RROOT | (suppress_errors ? SPAWN_NO_STDERR : 0 ),
             MOUNTPROG, MOUNTPROG, "-t", fsname, "-o", options, device, mntpt,
             NULL );
}

/**
 * Try to call do_mount() with every supported file system until a call
 * succeeds.
 * @param device device node to mount
 * @param mntpt desired mount point
 * @param async if not 0, the device will be mounted with 'async' (i. e. write
 *        caching)
 * @param noatime if not 0, the device will be mounted with 'noatime'
 * @param exec if not 0, the device will be mounted with 'exec'
 * @param force_write 1 for forced r/w, 0 for forced r/o, -1 for kernel default
 * @param iocharset charset to use for file name conversion; NULL for mount
 *        default
 * @param umask User specified umask (NULL for default)
 * @return last return value of do_mount (i. e. 0 on success, != 0 on error)
 */
int
do_mount_auto( const char* device, const char* mntpt, int async, 
        int noatime, int exec, int force_write, const char* iocharset, 
        const char* umask )
{
    const struct FS* fs;
    int nostderr = 1;
    int result = -1;

    for( fs = get_supported_fs(); fs->fsname; ++fs ) {
        /* don't suppress stderr if we try the last possible fs */
        if( (fs+1)->fsname == NULL )
            nostderr = 0;
        result = do_mount( device, mntpt, fs->fsname, async, noatime, exec,
                force_write, iocharset, umask, nostderr );
        if( result == 0 )
            break;

	/* sometimes VFAT fails when using iocharset; try again without */
	if( iocharset )
	    result = do_mount( device, mntpt, fs->fsname, async, noatime, exec,
                    force_write, NULL, umask, nostderr );
        if( result <= 0 )
            break;
    }
    return result;
}

/**
 * Lock given device.
 * param pid pid of program that holds the lock
 * @return 0 on success, -1 on error (message is printed in this case).
 */
int
do_lock( const char* device, pid_t pid )
{
    char lockdirpath[PATH_MAX];
    char lockfilepath[PATH_MAX];
    int pidlock;

    if( assert_dir( LOCKDIR, 0 ) )
        return -1;

    make_lockdir_name( device, lockdirpath, sizeof( lockdirpath ) );

    if( assert_dir( lockdirpath, 0 ) )
        return -1;

    /* only allow to create locks for existing pids, to prevent DOS attacks */
    if( !pid_exists( pid ) ) {
        fprintf( stderr, _("Error: cannot lock for pid %u, this process does not exist\n"), pid );
        return -1;
    }

    snprintf( lockfilepath, sizeof( lockfilepath ), "%s/%u", lockdirpath, (unsigned) pid );

    /* we need root for creating the pid lock file */
    get_root();
    get_groot();
    pidlock = open( lockfilepath, O_WRONLY|O_CREAT, 0644 );
    drop_groot();
    drop_root();

    if( pidlock < 0 ) {
        fprintf( stderr, _("Error: could not create pid lock file %s: %s\n"),
                lockfilepath, strerror( errno ) );
        return -1;
    }

    close( pidlock );

    return 0;
}

/**
 * Unlock given device.
 * param pid pid of program that holds the lock
 * @return 0 on success, -1 on error (message is printed in this case).
 */
int
do_unlock( const char* device, pid_t pid )
{
    char lockdirpath[PATH_MAX];
    char lockfilepath[PATH_MAX];
    int result;

    make_lockdir_name( device, lockdirpath, sizeof( lockdirpath ) );

    /* if no lock dir exists, device is not locked */
    if( !is_dir( lockdirpath ) )
        return 0;

    /* remove pid file first */
    if( pid ) {
        snprintf( lockfilepath, sizeof( lockfilepath ), "%s/%u", lockdirpath, (unsigned) pid );

        /* we need root for removing the pid lock file */
        get_root();
        result = unlink( lockfilepath );
        drop_root();

        if( result ) {
            /* ignore nonexistant lock files, but report other errors */
            if( errno != ENOENT ) {
                fprintf( stderr, _("Error: could not remove pid lock file %s: %s\n"),
                         lockfilepath, strerror( errno ) );
                return -1;
            }
        }
    }

    /* Try to rmdir the dir. If there are still files (pid-locks) in it, this
     * will fail. */
    get_root();
    result = rmdir( lockdirpath );
    drop_root();

    if( result ) {
        if( errno == ENOTEMPTY )
            return 0;
        perror( _("Error: do_unlock: could not remove lock directory") );
        return -1;
    }

    return 0;
}

/**
 * Remove stale pid locks from device's lock directory.
 */
void
clean_lock_dir( const char* device )
{
    char lockdirpath[PATH_MAX];
    char lockfilepath[PATH_MAX];
    DIR* lockdir;
    struct dirent* lockfile;

    make_lockdir_name( device, lockdirpath, sizeof( lockdirpath ) );

    debug( "Cleaning lock directory %s\n", lockdirpath );

    get_root();
    lockdir = opendir( lockdirpath );
    drop_root();

    if( !lockdir )
        return;

    while( ( lockfile = readdir( lockdir ) ) ) {
        if( !strcmp( lockfile->d_name, "." ) || !strcmp( lockfile->d_name, "..") )
            continue;
        
        debug( "  checking whether %s is alive\n", lockfile->d_name);

        if( !pid_exists( parse_unsigned( lockfile->d_name, E_INTERNAL ) ) ) {
            debug( "  %s is dead, removing lock file\n", lockfile->d_name);
            snprintf( lockfilepath, sizeof( lockfilepath ), "%s/%s",
                    lockdirpath, lockfile->d_name );
            get_root();
            unlink( lockfilepath );
            drop_root();
        }
    }

    /* remove the directory if it got empty */
    get_root();
    rmdir( lockdirpath );
    drop_root();
}

/**
 * Entry point.
 */
int
main( int argc, char** argv )
{
    char *devarg = NULL, *arg2 = NULL;
    char mntpt[MEDIA_STRING_SIZE];
    char device[PATH_MAX], mntptdev[PATH_MAX];
    char decrypted_device[PATH_MAX];
    const char* fstab_device;
    int is_real_path = 0;
    int async = 1;
    int noatime = 0;
    int exec = 0;
    int force_write = -1; /* 0: ro, 1: rw, -1: default */
    const char* use_fstype = NULL;
    const char* iocharset = NULL;
    const char* umask = NULL;
    const char* passphrase = NULL;
    int result;

    enum { MOUNT, LOCK, UNLOCK } mode = MOUNT;

    int  option;
    static struct option long_opts[] = {
        { "help", 0, NULL, 'h'},
        { "debug", 0, NULL, 'd'},
        { "lock", 0, NULL, 'l'},
        { "unlock", 0, NULL, 'L'},
        { "sync", 0, NULL, 's' },
        { "noatime", 0, NULL, 'A' },
        { "exec", 0, NULL, 'e' },
        { "type", 1, NULL, 't' },
        { "charset", 1, NULL, 'c' },
        { "umask", 1, NULL, 'u' },
        { "passphrase", 1, NULL, 'p' },
        { "read-only", 0, NULL, 'r' },
        { "read-write", 0, NULL, 'w' },
        { NULL, 0, NULL, 0}
    };

    /* initialize locale */
    setlocale( LC_ALL, "" );
    bindtextdomain( "pmount", NULL );
    textdomain( "pmount" );

    /* are we root? */
    if( geteuid() ) {
        fputs( _("Error: this program needs to be installed suid root\n"), stderr );
        return E_INTERNAL;
    }

    /* drop root privileges until we really need them (still available as saved uid) */
    seteuid( getuid() );

    /* parse command line options */
    do {
        switch( option = getopt_long( argc, argv, "+hdelLsArwt:c:u:", long_opts, NULL ) ) {
            case -1:  break;          /* end of arguments */
            case ':':
            case '?': return E_ARGS;  /* unknown argument */

            case 'h': usage( argv[0] ); return 0;

            case 'd': enable_debug = 1; break;

            case 'l': mode = LOCK; break;

            case 'L': mode = UNLOCK; break;

            case 's': async = 0; break;

            case 'A': noatime = 1; break;

            case 'e': exec = 1; break;

            case 't': use_fstype = optarg; break;

            case 'c': iocharset = optarg; break;

            case 'u': umask = optarg; break;

            case 'p': passphrase = optarg; break;

            case 'r': force_write = 0; break;

            case 'w': force_write = 1; break;

            default:
                fprintf( stderr, _("Internal error: getopt_long() returned unknown value\n") );
                return E_INTERNAL;
        }
    } while( option != -1 );

    /* determine device and second (label/pid) argument */
    if( optind < argc )
        devarg = argv[optind];

    if( optind+1 < argc )
        arg2 = argv[optind+1];

    /* check number of arguments */
    if( !devarg || ( mode != MOUNT && !arg2 ) || argc > optind+2 ) {
        usage( argv[0] );
        return E_ARGS;
    }

    /* if we got a mount point, convert it to a device */
    if( fstab_has_mntpt( "/etc/fstab", devarg, mntptdev, sizeof(mntptdev) ) ) {
        debug( "resolved mount point %s to device %s\n", devarg, mntptdev );
        devarg = mntptdev;
    }

    /* get real path, if possible */
    if( realpath( devarg, device ) ) {
        debug( "resolved %s to device %s\n", devarg, device );
        is_real_path = 1;
    } else {
        debug( "%s cannot be resolved to a proper device node\n", devarg );
        snprintf( device, sizeof( device ), "%s", devarg );
    }

    /* is the device already handled by fstab? We allow is_real_path == 0 here
     * to transparently mount thinks like NFS and SMB drives */
    fstab_device = fstab_has_device( "/etc/fstab", device, NULL, NULL );
    if( mode == MOUNT && fstab_device ) {
        if( arg2 )
            fprintf( stderr, _("Warning: device %s is already handled by /etc/fstab,"
                    " supplied label is ignored\n"), fstab_device );

        do_mount_fstab( fstab_device );
        return E_EXECMOUNT;
    }

    /* pmounted devices really have to be a proper local device */
    if( !is_real_path ) {
        perror( _("Error: could not determine real path of the device") );
        return E_DEVICE;
    }

    /* does the device start with DEVDIR? */
    if( strncmp( device, DEVDIR, sizeof( DEVDIR )-1 ) ) { 
        fprintf( stderr, _("Error: invalid device %s (must be in /dev/)\n"), device ); 
        return E_DEVICE;
    }

    switch( mode ) {
        case MOUNT:
            /* determine mount point name; note that we use devarg instead of
             * device to preserve symlink names (like '/dev/usbflash' instead
             * of '/dev/sda1') */
            if( make_mountpoint_name( devarg, arg2, mntpt, sizeof( mntpt ) ) )
                return E_MNTPT;

            /* if no charset was set explicitly, autodetect UTF-8 */
            if( !iocharset ) {
                const char* codeset;
                codeset = nl_langinfo( CODESET );

                debug( "no iocharset given, current locale encoding is %s\n", codeset );

                if( codeset && !strcmp( codeset, "UTF-8" ) ) {
                    debug( "locale encoding uses UTF-8, setting iocharset to 'utf8'\n" );
                    iocharset = "utf8";
                }
            }

            /* clean stale locks */
            clean_lock_dir( device );

            if( check_mount_policy( device, mntpt )  )
                return E_POLICY;

            /* check for encrypted device */
            enum decrypt_status decrypt = luks_decrypt( device,
                    decrypted_device, sizeof( decrypted_device ), passphrase,
                    force_write == 0 ? 1 : 0 ); 

            switch (decrypt) {
                case DECRYPT_FAILED:
                    fprintf( stderr, _("Error: could not decrypt device (wrong passphrase?)\n") );
                    exit( E_POLICY );
                case DECRYPT_EXISTS:
                    fprintf( stderr, _("Error: mapped device already exists\n") );
                    exit( E_POLICY );
                case DECRYPT_OK:
                case DECRYPT_NOTENCRYPTED:
                    break;
                default:
                    fprintf( stderr, "Internal error: unhandled decrypt_status %i\n", 
                        (int) decrypt);
                    exit( E_INTERNAL );
            }

            /* lock the mount directory */
            debug( "locking mount point directory\n" );
            if( lock_dir( mntpt ) < 0) {
                fprintf( stderr, _("Error: could not lock the mount directory. Another pmount is probably running for this mount point.\n"));
                exit( E_LOCKED );
            }
            debug( "mount point directory locked\n" );

            /* off we go */
            if( use_fstype )
                result = do_mount( decrypted_device, mntpt, use_fstype, async, noatime,
                        exec, force_write, iocharset, umask, 0 );
            else
                result = do_mount_auto( decrypted_device, mntpt, async, noatime, exec,
                        force_write, iocharset, umask ); 

            /* unlock the mount point again */
            debug( "unlocking mount point directory\n" );
            unlock_dir( mntpt );
            debug( "mount point directory unlocked\n" );

            if( result ) {
                if( decrypt == DECRYPT_OK )
                    luks_release( decrypted_device );

                /* mount failed, delete the mount point again */
                if( remove_pmount_mntpt( mntpt ) ) {
                    perror( _("Error: could not delete mount point") );
                    return -1;
                }
                return E_EXECMOUNT;
            }

            return 0;

        case LOCK:
            if( device_valid( device ) )
                if( do_lock( device, parse_unsigned( arg2, E_PID ) ) )
                    return E_INTERNAL;
            return 0;

        case UNLOCK:
            if( device_valid( device ) )
                if( do_unlock( device, parse_unsigned( arg2, E_PID ) ) )
                    return E_UNLOCK;
            return 0;
    }

    fprintf( stderr, _("Internal error: mode %i not handled.\n"), (int) mode );
    return E_INTERNAL;
}
